### ISSUE TITLE
DBI error: NOT NULL constraint failed: report_error.time

### ISSUE THREAD

**Issue by freddieleeman** [msimerson/mail-dmarc#158]
If a report could not be delivered (45x response) the dmarc_send_reports script returns the following error in verbose mode:

```
DBI error: NOT NULL constraint failed: report_error.time
Sending error query called by Mail::DMARC::Report::Store::SQL, 523
        INSERT INTO report_error (report_id, error ) VALUES (??)
        230, 454 4.7.1 <dmarc@example.com>: Relay access denied
 at /usr/local/share/perl/5.28.1/Mail/DMARC/Report/Store.pm line 17.
```

This issue occurs in the latest master build but also in v1.20191004

 - OS: Debian Buster
- PERL v5.28.1
- DSN: dbi:SQLite
- [Docker](https://github.com/freddieleeman/authmilter_docker)

**Comment by msimerson** [comment]
### report_error table definition

- MySQL: `time timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP`
- Postgres: `time timestamp NOT NULL DEFAULT now()`
- Sqlite: `time integer NOT NULL`

### Potential solutions

- always set the report_error.time value.
- ???

**Comment by freddieleeman** [comment]
Not much experience with Sqlite, but doesn't it too support timestamp and a default value just like MySQL?

`TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL`

**Comment by msimerson** [comment]
> doesn't it too support timestamp and a default value

Nope.

**Comment by msimerson** [comment]
I'm thinking this will do the needful:

```diff
diff --git a/lib/Mail/DMARC/Report/Store/SQL.pm b/lib/Mail/DMARC/Report/Store/SQL.pm
index d82b905..d7df9f0 100644
--- a/lib/Mail/DMARC/Report/Store/SQL.pm
+++ b/lib/Mail/DMARC/Report/Store/SQL.pm
@@ -437,7 +437,7 @@ sub insert_error {
 
     return $self->query(
         $self->grammar->insert_error(1),
-        [ $rid, $error ]
+        [ $rid, $error, time ]
     );
 }
 
diff --git a/lib/Mail/DMARC/Report/Store/SQL/Grammars/MySQL.pm b/lib/Mail/DMARC/Report/Store/SQL/Grammars/MySQL.pm
index 55c734d..d7854f2 100644
--- a/lib/Mail/DMARC/Report/Store/SQL/Grammars/MySQL.pm
+++ b/lib/Mail/DMARC/Report/Store/SQL/Grammars/MySQL.pm
@@ -211,7 +211,7 @@ sub insert_error {
     if ( $which == 0 ) {
         return 'UPDATE report SET end=? WHERE id=?';
     } else {
-        return 'INSERT INTO report_error (report_id, error) VALUES (?,?)';
+        return 'INSERT INTO report_error (report_id, error, time) VALUES (??)';
     }
 }
 
diff --git a/lib/Mail/DMARC/Report/Store/SQL/Grammars/PostgreSQL.pm b/lib/Mail/DMARC/Report/Store/SQL/Grammars/PostgreSQL.pm
index 1810785..7a8f631 100644
--- a/lib/Mail/DMARC/Report/Store/SQL/Grammars/PostgreSQL.pm
+++ b/lib/Mail/DMARC/Report/Store/SQL/Grammars/PostgreSQL.pm
@@ -207,7 +207,7 @@ sub insert_error {
     if ( $which == 0 ) {
         return 'UPDATE "report" SET "end"=? WHERE "id"=?';
     } else {
-        return 'INSERT INTO "report_error" ("report_id", "error") VALUES (?,?)';
+        return 'INSERT INTO "report_error" ("report_id", "error", "time") VALUES (??)';
     }
 }
 
diff --git a/lib/Mail/DMARC/Report/Store/SQL/Grammars/SQLite.pm b/lib/Mail/DMARC/Report/Store/SQL/Grammars/SQLite.pm
index 569a5c7..5b4b04c 100644
--- a/lib/Mail/DMARC/Report/Store/SQL/Grammars/SQLite.pm
+++ b/lib/Mail/DMARC/Report/Store/SQL/Grammars/SQLite.pm
@@ -205,7 +205,7 @@ sub insert_error {
     if ( $which == 0 ) {
         return 'UPDATE report SET end=? WHERE id=?';
     } else {
-        return 'INSERT INTO report_error (report_id, error) VALUES (?,?)';
+        return 'INSERT INTO report_error (report_id, error, time) VALUES (??)';
     }
 }
 
```

**Comment by freddieleeman** [comment]
```
sqlite> CREATE TABLE test_table (
   ...>     id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
   ...>     name VARCHAR(64),
   ...>     time TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
   ...> );
sqlite> INSERT INTO test_table(name) VALUES('test');
sqlite> SELECT * FROM test_table;
1|test|2019-12-26 07:32:27
```

**Comment by msimerson** [comment]
Okay, I learned something. As of v 3.1, sqlite supports a 'timestamp' field type. Sort of...

> SQLite does not support built-in date and/or time storage class. Instead, it leverages some built-in date and time functions to use other storage classes such as TEXT, REAL, or INTEGER for storing the date and time values.

So sqlite added some syntactic sugar and store a timestamp as a text string. And they support default values now. 🤭 I have no idea what sqlite versions are in the wild nor how common > v3.1 is. Requiring a newish version wouldn't be terrible.

It appears that `report_error.time` is only a human reference. I don't see it being selected or used anywhere.

If we change the sqlite DB definition to a timestamp/text, then people upgrading need a conversion/upgrade script to convert any stored INT values to TEXT.  But then, if it hasn't been working for sqlite users, maybe that's not necessary.

I did it right in the `report` table, the start and end times are `INT`s, so no DB timestamp handling to worry about. I'm going to cogitate on this.
